README.md
# Rfbeam

![Gem](https://img.shields.io/gem/v/rfbeam?color=green&label=version)
![Ruby](https://img.shields.io/static/v1?message=Ruby&color=red&logo=Ruby&logoColor=FFFFFF&label=v3.2.1)
![Ruby](https://img.shields.io/gitlab/license/robcarruthers/rfbeam?color=orange)

RfBeam is a simple, high-level interface for the RFBeam radar modules.
The user can query process and raw detection data and set the radar parameters for the sensor.

At this stage it only works on Linux and Mac with the K-LD7 module.

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add rfbeam

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install rfbeam

## Usage

The RfBeam class will return the path of any connected modules

    RfBeam.connected

Simple pass the path and baude rate to initialise a new radar object

    RfBeam::Kld7.new("/dev/ttyUSB0", 115200) do |radar|
        puts radar.config
    end

## RfBeam::Kld7 API

### detection?

Returns true if the module has a valid detection.

    radar.detection?
    => true

### ddat

Returns an array with the current detection values

    radar.ddat
    => ["DDAT", 6, 0, 0, 0, 0, 0, 0]

### tdat

Returns a hash with the current tracked object values

    radar.tdat
    => {:dist=>68, :speed=>196, :angle=>469, :mag=>6303}

### grps

Returns the Parameter settings, values map to setting as detailed in the device datasheet.

    radar.grps
    => ["RPST", 42, "K-LD7_APP-RFB-0103", 1, 1, 1, 30, 0, 2, 0, 50, -90, 90, 0, 100, 2, 10, 0, 50, 0, 1, 2, 1, 0, 4]

### config

Returns a formatted String of all parameter settings. The only way to read parameter settings is with config

    radar.config

    Software Version: K-LD7_APP-RFB-0103
    Base Frequency: Low
    Maximum Speed: 100km/h
    Maximum Range: 100m
    Threshold Offset: 30db
    Tracking Filter Type: Long Visibility
    Vibration Suppression: 16
    Minimum Detection Distance: 0%
    Maximum Detection Distance: 100%
    Minimum Detection Angle: -10°
    Maximum Detection Angle: 90°
    Minimum Detection Speed: 0%
    Maximum Detection Speed: 100%
    Detection Direction: Both
    Range Threshold: 10%
    Angle Threshold: 0°
    Speed Threshold: 50%
    Digital Output 1: Direction
    Digital Output 2: Angle
    Digital Output 3: Range
    Hold Time: 1s
    Micro Detection Trigger: Off
    Micro Detection Sensativity: 4

## Parameter setters

### Base Frequency

0 = low, 1 = middle (default), 2 = high

alias :rbfr

```ruby
set_base_frequency(1)
```

### Maximum Speed

0 = 12.5km/h, 1 = 25km/h (default), 2 = 50km/h, 3 = 100km/h, alias :rspi

```ruby
set_max_speed(1)
```

### Maximum Range

0 = 5m, 1 = 10m (default), 2 = 30m, 3 = 100m, alias :rrai

```ruby
set_max_range(1)
```

### Threshold Offset

10 - 60db, (default = 30), alias :thof

```ruby
set_threshold_offset(30)
```

### Tracking filter type

0 = Standard (Default), 1 = Fast Tracking, 2 = Long visibility, alias :trtf

```ruby
set_tracking_filter(0)
```

### Vibration suppression

0 - 16, 0 = No Suppression, 16 = High Suppression, default = 2, alias :visu

```ruby
set_vibration_suppression(2)
```

### Minimum Detection distance

0 - 100% of Range setting, default = 0, alias :mira

```ruby
set_min_detection_distance(0)
```

### Maximum Detection distance

0 - 100% of Range setting, default = 50, alias :mara

```ruby
set_max_detection_distance(50)
```

### Minimum Detection Angle

-90° - 90°, default = -90, alias :mian

```ruby
set_min_detection_angle(-90
```

### Maximum Detection Angle

-90° - 90°, default = 90, alias :maan

```ruby
set_min_detection_angle(90)
```

### Minimum Detection Speed

0 - 100% of Speed setting, default = 0, alias :misp

```ruby
set_min_detection_speed(0)
```

### Maximum Detection Speed

0 - 100% of Speed setting, default = 100, alias :masp

```ruby
set_max_detection_speed(100)
```

### Detection Direction

0 = Receding, 1 = Approaching, 2 = Both (default), alias :dedi

```ruby
 set_detection_direction(2)
```

### Range Threshold

0 - 100% of Range setting, default = 10, alias :rath

```ruby
 set_range_threshold(10)
```

### Angle Threshold

-90° to 90°, default = 0, alias :anth

```ruby
 set_range_threshold(0)
```

### Speed Threshold

0 - 100% of speed setting, default = 50, alias :spth

```ruby
 set_angle_threshold(50)
```

### Digital output 1

0 = Direction, 1 = Angle, 2 = Range, 3 = Speed, 4 = Micro Detection, default = 0

alias :dig1, :set_dio_1

```ruby
set_dio_1(0)
```

### Digital output 2

0 = Direction, 1 = Angle, 2 = Range, 3 = Speed, 4 = Micro Detection, default = 1

alias :dig2, :set_dio_2

```ruby
set_dio_2(1)
```

### Digital output 3

0 = Direction, 1 = Angle, 2 = Range, 3 = Speed, 4 = Micro Detection, default = 2

alias :dig3, :set_dio_3

```ruby
set_dio_3(2)
```

### Hold Time

1 - 7200s, default = 1

alias :hold

```ruby
set_hold_time(1)
```

### Micro Detection retrigger

0 = Off (default), 1 = Retrigger

alias: :mide

```ruby
set_micro_detection_retrigger(0)
```

### Micro Detection sensitivity

0 - 9, 0 = Min, 9 = Max, default = 4

alias: :mids

```ruby
set_micro_detection_sensitivty(4)
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/rfbeam.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## Unreleased ([05e7324..9153d75](https://github.com/robcarruthers/rfbeam/compare/05e7324..9153d75))
#### Refactoring
- **(rubocop)** Fix offences - ([9153d75](https://github.com/robcarruthers/rfbeam/commit/9153d75248fa4b83a1a89bdd67a7286ce7ea4749)) - [@robcarruthers](https://github.com/robcarruthers)
- **(rubocop)** Fix errors - ([57359b8](https://github.com/robcarruthers/rfbeam/commit/57359b85c0f2262fc612b85856f7c90e5c5c6927)) - [@robcarruthers](https://github.com/robcarruthers)
- Refine response code logic - ([e6a88c5](https://github.com/robcarruthers/rfbeam/commit/e6a88c54abe7ccd5bf4b5696a3ac8255417abfd2)) - [@robcarruthers](https://github.com/robcarruthers)
#### Tests
- **(kld7)** Start adding tests - ([8b84680](https://github.com/robcarruthers/rfbeam/commit/8b846801bc65e55ad7092e0d9065287749ffb55b)) - [@robcarruthers](https://github.com/robcarruthers)

- - -
## [v0.5.8](https://github.com/robcarruthers/rfbeam/compare/v0.5.7..v0.5.8) - 2023-05-15
#### Refactoring
- **(rubocop)** Fix offences - ([9153d75](https://github.com/robcarruthers/rfbeam/commit/9153d75248fa4b83a1a89bdd67a7286ce7ea4749)) - [@robcarruthers](https://github.com/robcarruthers)
- **(rubocop)** Fix errors - ([57359b8](https://github.com/robcarruthers/rfbeam/commit/57359b85c0f2262fc612b85856f7c90e5c5c6927)) - [@robcarruthers](https://github.com/robcarruthers)
- Refine response code logic - ([e6a88c5](https://github.com/robcarruthers/rfbeam/commit/e6a88c54abe7ccd5bf4b5696a3ac8255417abfd2)) - [@robcarruthers](https://github.com/robcarruthers)
#### Tests
- **(kld7)** Add specs - ([96c8e93](https://github.com/robcarruthers/rfbeam/commit/96c8e93aae186e7ab14a4bad27efc50d11003c57)) - [@robcarruthers](https://github.com/robcarruthers)
- **(kld7)** Start adding tests - ([8b84680](https://github.com/robcarruthers/rfbeam/commit/8b846801bc65e55ad7092e0d9065287749ffb55b)) - [@robcarruthers](https://github.com/robcarruthers)

- - -


## [v0.5.7](https://github.com/robcarruthers/rfbeam/compare/v0.5.1..v0.5.7) - 2023-05-15
#### Miscellaneous Tasks
- **(changelog)** Update - ([7299d86](https://github.com/robcarruthers/rfbeam/commit/7299d86e4c1948b9f157c8064b5559056a48a162)) - [@robcarruthers](https://github.com/robcarruthers)
- **(changelog)** Update cog config - ([8de4b62](https://github.com/robcarruthers/rfbeam/commit/8de4b62455a3ba2c917149e3586624167156c09f)) - [@robcarruthers](https://github.com/robcarruthers)
- **(changelog)** Config git-cliff - ([e212231](https://github.com/robcarruthers/rfbeam/commit/e212231caa1170a2c2e5d000550bb4623516b865)) - [@robcarruthers](https://github.com/robcarruthers)
- **(changelog)** Update - ([2f2e8e1](https://github.com/robcarruthers/rfbeam/commit/2f2e8e1f577e210bb2e245e38cbaa24203c33a03)) - [@robcarruthers](https://github.com/robcarruthers)
- **(version)** v0.5.7 - ([05e7324](https://github.com/robcarruthers/rfbeam/commit/05e73242fa1c7473aa36465fde50d84850ff5f3a)) - [@robcarruthers](https://github.com/robcarruthers)
- **(version)** v0.5.6 - ([64af4cb](https://github.com/robcarruthers/rfbeam/commit/64af4cbf2b18ead1c42728d9af6a9577211eeb8f)) - [@robcarruthers](https://github.com/robcarruthers)
- **(version)** v0.5.5 - ([c0f6688](https://github.com/robcarruthers/rfbeam/commit/c0f66887decb8444491025390d3a69f593b81631)) - [@robcarruthers](https://github.com/robcarruthers)
- **(version)** v0.5.4 - ([4e12296](https://github.com/robcarruthers/rfbeam/commit/4e12296e994ba07815aaeba4ca119a13223f71c8)) - [@robcarruthers](https://github.com/robcarruthers)
- **(version)** v0.5.3 - ([560acf7](https://github.com/robcarruthers/rfbeam/commit/560acf78d1bc792a94a18de5f649b2a056a438b9)) - [@robcarruthers](https://github.com/robcarruthers)
- v0.5.2 - ([aae010c](https://github.com/robcarruthers/rfbeam/commit/aae010c451f9c9d6cea274fd0dd2216278507f41)) - [@robcarruthers](https://github.com/robcarruthers)

- - -

## [v0.5.1](https://github.com/robcarruthers/rfbeam/compare/v0.4.0..v0.5.1) - 2023-05-15
#### Bug Fixes
- **(cli)** Update logger output - ([f80b78e](https://github.com/robcarruthers/rfbeam/commit/f80b78e289cc737a2ddae33df30414f3c1933beb)) - [@robcarruthers](https://github.com/robcarruthers)
#### Miscellaneous Tasks
- **(changelog)** Update - ([92fff48](https://github.com/robcarruthers/rfbeam/commit/92fff489f58e3c4a1c104fea634d78f72857aea8)) - [@robcarruthers](https://github.com/robcarruthers)
- **(version)** v0.5.1 - ([530b254](https://github.com/robcarruthers/rfbeam/commit/530b254f86477aaf75efebf773b12bf6abfe164a)) - [@robcarruthers](https://github.com/robcarruthers)
#### Tests
- Setup minitest - ([e8ad2a8](https://github.com/robcarruthers/rfbeam/commit/e8ad2a8233ca653bd21871a91b668388bc63416c)) - [@robcarruthers](https://github.com/robcarruthers)

- - -

## [v0.4.0](https://github.com/robcarruthers/rfbeam/compare/87bcfe5bac1f73a5a0ba4445266e828723ee2289..v0.4.0) - 2023-05-15




lib/rfbeam/kld7/constants.rb
# frozen_string_literal: true
# rubocop:disable all

module RfBeam
  module KLD7
    # All supported Serial port baude rates
    BAUDE_RATES = { 0 => 115_200, 1 => 460_800, 2 => 921_600, 3 => 2_000_000, 4 => 3_000_000 }.freeze

    # 'INIT' command response codes
    RESP_CODES = {
      0 => 'OK',
      1 => 'Unknown command',
      2 => 'Invalid parameter value',
      3 => 'Invalid RPST version',
      4 => 'Uart error (parity, framing, noise)',
      5 => 'Sensor busy',
      6 => 'Timeout error'
    }.freeze

    # Delays are determined empirically and may need adjusting with baude rate
    RESPONSE_DELAY = 0.05
    MEASUREMENT_DELAY = 0.15

    # 'GNFD' command types
    FRAME_DATA_TYPES = { disabled: 0x00, radc: 0x01, rfft: 0x02, pdat: 0x04, tdat: 0x08, ddat: 0x10, done: 0x20 }.freeze

    # The angle, direction, range and speed flags are only valid if the detection flag is 1.
    DETECTION_FLAGS = {
      detection: ['No Detection', 'Detection'],
      micro_detection: ['No Detection', 'Detection'],
      angle: %w[Left Right],
      direction: %w[Receding Approaching],
      range: %w[Far Near],
      speed: %w[Low High]
    }.freeze

    Param =
      Data.define(:name, :grps_index, :description, :default, :units, :values) do |_param|
        def initialize(name:, grps_index:, description: nil, default: nil, units: nil, values: [])
          super(name:, grps_index:, description:, default:, units:, values:)
        end
      end

    RADAR_PARAMETERS = {
      sw_version: Param.new(name: 'Software Version', grps_index: 2, default: 'K-LD7_APP-RFB-XXXX'),
      base_frequency:
        Param.new(
          name: 'Base Frequency',
          grps_index: 3,
          description: '0 = Low, 1 = Middle, 2 = High',
          default: 1,
          values: %w[Low Middle High]
        ),
      max_speed:
        Param.new(
          name: 'Maximum Speed',
          grps_index: 4,
          description: '0 = 12km/h, 1 = 25km/h, 2 = 50km/h, 3 = 100km/h',
          default: 1,
          units: 'km/h',
          values: %w[12.5 25 50 100]
        ),
      max_range:
        Param.new(
          name: 'Maximum Range',
          grps_index: 5,
          description: '0 = 5m, 1 = 10m, 2 = 30m, 3 = 100m',
          default: 1,
          values: %w[5m 10m 30m 100m]
        ),
      threshold_offset:
        Param.new(name: 'Threshold Offset', grps_index: 6, description: '10db - 60db', default: 30, units: 'db'),
      tracking_filter:
        Param.new(
          name: 'Tracking Filter Type',
          grps_index: 7,
          description: '0 = Standard, 2 = Fast Detection, 3 = Long Visibility',
          default: 0,
          values: ['standard', 'Fast Detection', 'Long Visibility']
        ),
      vibration_suppression:
        Param.new(
          name: 'Vibration Suppression',
          grps_index: 8,
          description: '0-16, 0 = No Suppression, 16 = High Suppression',
          default: 2
        ),
      min_detection_distance:
        Param.new(
          name: 'Minimum Detection Distance',
          grps_index: 9,
          description: '0 - 100% of range setting',
          default: 0,
          units: '%'
        ),
      max_detection_distance:
        Param.new(
          name: 'Maximum Detection Distance',
          grps_index: 10,
          description: '0 - 100% of range setting',
          default: 50,
          units: '%'
        ),
      min_detection_angle:
        Param.new(name: 'Minimum Detection Angle', grps_index: 11, description: '-90° - 90°', default: -90, units: '°'),
      max_detection_angle:
        Param.new(name: 'Maximum Detection Angle', grps_index: 12, description: '-90° - 90°', default: 90, units: '°'),
      min_detection_speed:
        Param.new(
          name: 'Minimum Detection Speed',
          grps_index: 13,
          description: '0 - 100% of speed setting',
          default: 0,
          units: '%'
        ),
      max_detection_speed:
        Param.new(
          name: 'Maximum Detection Speed',
          grps_index: 14,
          description: '0 - 100% of speed setting',
          default: 100,
          units: '%'
        ),
      detection_direction:
        Param.new(
          name: 'Detection Direction',
          grps_index: 15,
          description: '0 = Receding, 1 = Approaching, 2 = Both',
          default: 2,
          values: %w[Receding Approaching Both]
        ),
      range_threshold:
        Param.new(
          name: 'Range Threshold',
          grps_index: 16,
          description: '0 - 100% of range setting',
          default: 10,
          units: '%'
        ),
      angle_threshold:
        Param.new(name: 'Angle Threshold', grps_index: 17, description: '-90° - 90°', default: 0, units: '°'),
      speed_threshold:
        Param.new(
          name: 'Speed Threshold',
          grps_index: 18,
          description: '0 - 100% of speed setting',
          default: 50,
          units: '%'
        ),
      digital_output1:
        Param.new(
          name: 'Digital Output 1',
          grps_index: 19,
          description: '0 = Direction, 1 = Angle, 2 = Range, 3 = Speed, 4 = Micro Detection',
          default: 0,
          values: %w[Direction Angle Range Speed Micro]
        ),
      digital_output2:
        Param.new(
          name: 'Digital Output 2',
          grps_index: 20,
          description: '0 = Direction, 1 = Angle, 2 = Range, 3 = Speed, 4 = Micro Detection',
          default: 1,
          values: %w[Direction Angle Range Speed Micro]
        ),
      digital_output3:
        Param.new(
          name: 'Digital Output 3',
          grps_index: 21,
          description: '0 = Direction, 1 = Angle, 2 = Range, 3 = Speed, 4 = Micro Detection',
          default: 2,
          values: %w[Direction Angle Range Speed Micro]
        ),
      hold_time: Param.new(name: 'Hold Time', grps_index: 22, description: '1 - 7200s', default: 1, units: 's'),
      micro_detection_retrigger:
        Param.new(
          name: 'Micro Detection Trigger',
          grps_index: 23,
          description: '0 = Off, 1 = Retrigger',
          default: 0,
          values: %w[Off Retrigger]
        ),
      micro_detection_sensativity:
        Param.new(
          name: 'Micro Detection Sensativity',
          grps_index: 24,
          description: '0 - 9, 0 = Min, 9 = Max',
          default: 4
        )
    }.freeze
  end
end
